File: 3-2.py
#!/usr/bin/env python

import math

readings = [r.strip() for r in open('3.input').readlines()]
word_len = len(readings[0])

oxygen = readings.copy()
co2 = readings.copy()

for bit in range(word_len):
    if len(oxygen) > 1:
        ones = 0
        most_common = 0
        for o in oxygen:
            ones += int(o[bit])
        if ones >= math.ceil(len(oxygen) / 2):
            most_common = 1
        oxygen = [o for o in oxygen if int(o[bit]) != most_common]

    if len(co2) > 1:
        ones = 0
        most_common = 0
        for c in co2:
            ones += int(c[bit])
        if ones >= math.ceil(len(co2) / 2):
            most_common = 1
        co2 = [c for c in co2 if int(c[bit]) == most_common]

life_support = int(oxygen[0], 2) * int(co2[0], 2)

print(life_support)

File: 3-2.rb
#!/usr/bin/env ruby

readings = File.read('3.input').lines.map(&:strip)
word_len = readings[0].size

oxygen = readings.dup
co2 = readings.dup

0.upto(word_len - 1) do |bit|
  if oxygen.size > 1
    ones = 0
    most_common = 0
    oxygen.each do |o|
      ones += o[bit].to_i
    end
    most_common = 1 if ones >= (oxygen.size / 2.0).ceil
    oxygen.delete_if { |o| o[bit].to_i != most_common }
  end

  if co2.size > 1
    ones = 0
    most_common = 0
    co2.each do |c|
      ones += c[bit].to_i
    end
    most_common = 1 if ones >= (co2.size / 2.0).ceil
    co2.delete_if { |c| c[bit].to_i == most_common }
  end
end

life_support = oxygen[0].to_i(2) * co2[0].to_i(2)

print life_support, "\n"
